src/engine/WordIndexScan.cpp

```diff
@@ -0,0 +1,94 @@
+#include "engine/WordIndexScan.h"
+
+// _____________________________________________________________________________
+WordIndexScan::WordIndexScan(QueryExecutionContext* qec,
+                             SetOfVariables variables, Variable cvar,
+                             string word)
+    : Operation(qec),
+      variables_(std::move(variables)),
+      cvar_(std::move(cvar)),
+      word_(word) {
+  if (word_.ends_with('*')) {
+    isPrefix_ = true;
+  }
+}
+
+// _____________________________________________________________________________
+ResultTable WordIndexScan::computeResult() {
+  LOG(INFO) << "words compute: " << word_ << endl;
+  IdTable idTable{getExecutionContext()->getAllocator()};
+  idTable.setNumColumns(1 + variables_.size() + isPrefix_);
+  Index::WordEntityPostings wep =
+      getExecutionContext()->getIndex().getEntityPostingsForTerm(word_);
+  idTable.resize(wep.cids_.size());
+  int i = 0;
+  decltype(auto) cidColumn = idTable.getColumn(i++);
+  std::ranges::transform(wep.cids_, cidColumn.begin(),
+                         &Id::makeFromTextRecordIndex);
+  if (isPrefix_) {
+    decltype(auto) widColumn = idTable.getColumn(i++);
+    std::ranges::transform(wep.wids_[0], widColumn.begin(), [](WordIndex id) {
+      return Id::makeFromWordVocabIndex(WordVocabIndex::make(id));
+    });
+  }
+  decltype(auto) scoreColumn = idTable.getColumn(i++);
+  std::ranges::transform(wep.scores_, scoreColumn.begin(), &Id::makeFromInt);
+  if (variables_.size() - 1 == 1) {
+    decltype(auto) eidColumn = idTable.getColumn(i++);
+    std::ranges::copy(wep.eids_.begin(), wep.eids_.end(), eidColumn.begin());
+  } else if (variables_.size() - 1 >= 2) {
+    // TODO: implement mult var case
+  }
```

Comment:
I think we never even need an `entity` column for this class.
Either we have a scan for a word, then we get a table with `TEXT | WORD(ifPrefix)|`.
Then we also need  textIndexEntityScan(word, variable) that returns `TEXT | Entity ` .
The rest is then done by the queryPlanner via Joins. 
So basically you also have one scan for each variables, but those only get the entity posings for the matching prefixes. 
Then in this class above, you only call `getWordPostingsForTerm`.

This second class (EntityScanForTextPrefix) is very much the same as this class then.

Reply:
Yes, I am beginning to think that splitting them up is a good idea. At the moment I think it would be best, if we had a `wordScan-Operation`, which gets called once for each word in `?text ql:contains-word "word1 ... wordN"`. So each of the Operations returns a table `?text | word1 (ifPrefix)` ... `?text | wordN (ifPrefix)`.
And a `entityScan-Operation`, which gets called for each `?text ql:contains-entity ?var1`. So each Operation returns a table `?text | ?var1`.
And then the join could do the rest.